src/spi.rs

```diff
@@ -0,0 +1,422 @@
+                            .set_bit()
+                            .ssi()
+                            .set_bit()
+                            .rxonly()
+                            .clear_bit()
+                            .dff()
+                            .clear_bit()
+                            .bidimode()
+                            .clear_bit()
+                            .spe()
+                            .set_bit()
+                    });
+
+                    Spi { spi, pins }
+                }
+
+                pub fn free(self) -> ($SPIX, PINS) {
+                    (self.spi, self.pins)
+                }
+            }
+
+            impl<PINS> hal::spi::FullDuplex<u8> for Spi<$SPIX, PINS> {
+                type Error = Error;
+
+                fn read(&mut self) -> nb::Result<u8, Error> {
+                    let sr = self.spi.sr.read();
+
+                    Err(if sr.ovr().bit_is_set() {
+                        nb::Error::Other(Error::Overrun)
+                    } else if sr.modf().bit_is_set() {
+                        nb::Error::Other(Error::ModeFault)
+                    } else if sr.crcerr().bit_is_set() {
+                        nb::Error::Other(Error::Crc)
+                    } else if sr.rxne().bit_is_set() {
+                        // NOTE(read_volatile) read only 1 byte (the svd2rust API only allows
+                        // reading a half-word)
+                        return Ok(unsafe {
+                            ptr::read_volatile(&self.spi.dr as *const _ as *const u8)
+                        });
+                    } else {
+                        nb::Error::WouldBlock
+                    })
+                }
+
+                fn send(&mut self, byte: u8) -> nb::Result<(), Error> {
+                    let sr = self.spi.sr.read();
+
+                    Err(if sr.ovr().bit_is_set() {
+                        nb::Error::Other(Error::Overrun)
+                    } else if sr.modf().bit_is_set() {
+                        nb::Error::Other(Error::ModeFault)
+                    } else if sr.crcerr().bit_is_set() {
+                        nb::Error::Other(Error::Crc)
+                    } else if sr.txe().bit_is_set() {
+                        // NOTE(write_volatile) see note above
+                        unsafe { ptr::write_volatile(&self.spi.dr as *const _ as *mut u8, byte) }
+                        return Ok(());
+                    } else {
+                        nb::Error::WouldBlock
+                    })
+                }
+
+            }
+
+            impl<PINS> ::hal::blocking::spi::transfer::Default<u8> for Spi<$SPIX, PINS> {}
+
+            impl<PINS> ::hal::blocking::spi::write::Default<u8> for Spi<$SPIX, PINS> {}
+        )+
+    }
+}
+
+hal! {
+    SPI1: (spi1, apb2enr, spi1en, pclk2),
+    SPI2: (spi2, apb1enr, spi2en, pclk1),
+    SPI3: (spi3, apb1enr, spi3en, pclk1),
+}
+
+// TODO(psycowithespn) Enable all of these on the next version of stm32-rs.
```

Comment:
Please remove this block for now. I don't want random todos hanging around in the code.